apis/v1alpha1/types.go

```diff
@@ -114,3 +114,98 @@ type ProviderConfigUsageList struct {
 	metav1.ListMeta `json:"metadata,omitempty"`
 	Items           []ProviderConfigUsage `json:"items"`
 }
+
+// +kubebuilder:object:root=true
+
+// A ObservedObjectCollection is a provider Kubernetes API type
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="KIND",type="string",JSONPath=".spec.kind"
+// +kubebuilder:printcolumn:name="APIVERSION",type="string",JSONPath=".spec.apiVersion",priority=1
+// +kubebuilder:printcolumn:name="PROVIDERCONFIG",type="string",JSONPath=".spec.providerConfigRef.name"
+// +kubebuilder:printcolumn:name="SYNCED",type="string",JSONPath=".status.conditions[?(@.type=='Synced')].status"
+// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".status.conditions[?(@.type=='Ready')].status"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Cluster,categories={crossplane,managed,kubernetes}
+type ObservedObjectCollection struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              ObservedObjectCollectionSpec   `json:"spec"`
+	Status            ObservedObjectCollectionStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ObservedObjectCollectionList contains a list of ObservedObjectCollection
+type ObservedObjectCollectionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ObservedObjectCollection `json:"items"`
+}
+
+// ObservedObjectCollectionSpec defines the desired state of ObservedObjectCollection
+type ObservedObjectCollectionSpec struct {
+
+	// APIVersion of objects that should be matched by the selector
+	// +kubebuilder:validation:MinLength:=1
+	APIVersion string `json:"apiVersion"`
+
+	// Kind of objects that should be matched by the selector
+	// +kubebuilder:validation:MinLength:=1
+	Kind string `json:"kind"`
+
+	// Namespace where to look for objects.
+	// If omitted, search is performed across all namespaces.
+	// For cluster-scoped objects, omit it.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Selector defines the criteria for including objects into the collection
+	Selector metav1.LabelSelector `json:"selector"`
```

Comment:
`match`?

Reply:
Not necessarily, I see the discussion there is around the general approach on how to tackle the problem at hand rather than the final shape of API.

Given this will be our API (and it is possible that some future examples will follow the same pattern), it is worth putting more thought before committing.

Reply:
Yeah, I like the idea of nesting the fields that specify what resources to observe.

Of all the names I've seen so far, `spec.observe` sounds best to me. Maybe `spec.observeObjects`?

Reply:
Thanks Nic for the suggestino - we went with `spec.observedObjects`.

Reply:
Are you sure we want `spec.observedObjects` (i.e. observed, not observe)?

Observed is the past tense of observe. So to me `ObservedObjectCollection` makes sense as the kind, because once you declare its existence what you have is a collection of observed objects - an "observed object collection".

Past tense makes less sense to me in the context of the `spec` though, since that's where I'm telling the collection what to do - "observe these objects". So to me `spec.observeObjects` is more intuitive.

I don't feel strongly about this though, and I could just be overthinking it. 🙂